File: 2405-optimal-partition-of-string/2405-optimal-partition-of-string.c
#include <stdio.h>
#include <stdbool.h>

/**
 * @brief Finds the minimum number of substrings with no repeating characters.
 *
 * This function partitions the input string into the smallest number of substrings
 * such that no substring contains repeated characters.
 *
 * @param s (const char*): The input string to partition.
 *
 * @return (int): The number of substrings created.
 */
int partitionString(char* s) {
    bool seen_chars[26] = {false};  /* Boolean array to track characters in the current substring */
    int count = 0;                  /* Counter for the number of substrings created */
    int i = 0;                      /* Index for iterating over the string */

    while (s[i] != '\0') {
        int idx = s[i] - 'a';  /* Calculate the index for the current character (assuming only lowercase letters) */
        
        if (seen_chars[idx]) {
            /* If the character is already in the current substring,
               start a new substring and increment the count */
            count++;
            for (int j = 0; j < 26; j++) {
                seen_chars[j] = false;  /* Clear the boolean array for the new substring */
            }
        }
        
        /* Mark the current character as seen */
        seen_chars[idx] = true;
        
        /* Move to the next character */
        i++;
    }
    
    /* Return the number of substrings (including the last one) */
    return count + 1;  /* +1 to account for the last substring */
}

File: 2405-optimal-partition-of-string/2405-optimal-partition-of-string.py
class Solution:
    def partitionString(self, s: str) -> int:
        """
        Partition the input string into the smallest number of substrings 
        such that no substring contains repeated characters.

        Parameters:
        s (str): The input string to partition.

        Returns:
        int: The number of substrings created.
        """
        seen = 0  # Bitmask to track characters in the current substring
        count = 0  # Counter for the number of substrings created

        for ch in s:
            bit_position = ord(ch) - ord('a')  # Calculate bit position for the character
            if seen & (1 << bit_position):
                # If character already seen, increment the partition count and reset the bitmask
                count += 1
                seen = 0
            
            # Mark the character as seen by setting the corresponding bit
            seen |= (1 << bit_position)

        # Return the total count of substrings, including the last one
        return count + 1
